src/test/osd/RadosModel.h

```diff
@@ -1948,6 +1966,197 @@ class CopyFromOp : public TestOp {
+      if (!tgt_value.deleted())
+	context->update_object_full(oid_tgt, tgt_value);
+    }
+
+    if (src_value.version != 0)
+      op.assert_version(src_value.version);
+    op.set_redirect(context->prefix+oid_tgt, context->io_ctx, tgt_value.version);
+
+    pair<TestOp*, TestOp::CallbackInfo*> *cb_arg =
+      new pair<TestOp*, TestOp::CallbackInfo*>(this,
+					       new TestOp::CallbackInfo(0));
+    comp = context->rados.aio_create_completion((void*) cb_arg, NULL,
+						&write_callback);
+    context->io_ctx.aio_operate(context->prefix+oid, comp, &op,
+				librados::OPERATION_ORDER_READS_WRITES);
+
+    pair<TestOp*, TestOp::CallbackInfo*> *read_cb_arg =
+     new pair<TestOp*, TestOp::CallbackInfo*>(this,
+     					 new TestOp::CallbackInfo(1));
+    comp_stat_read = context->rados.aio_create_completion((void*) read_cb_arg, NULL, &write_callback);
+    rd_op.stat(NULL, NULL, NULL);
+    context->io_ctx.aio_operate(context->prefix+oid, comp_stat_read, &rd_op,
+     			  librados::OPERATION_ORDER_READS_WRITES |
+     			  librados::OPERATION_IGNORE_REDIRECT,
+     			  NULL);
+  }
+
+  void _finish(CallbackInfo *info) override
+  {
+    Mutex::Locker l(context->state_lock);
+
+    if (info->id == 0) {
+      assert(comp->is_complete());
+      cout << num << ":  finishing set_redirect to oid " << oid << std::endl;
+      if ((r = comp->get_return_value())) {
+	if (r == -ENOENT && src_value.deleted()) {
+	  cout << num << ":  got expected ENOENT (src dne)" << std::endl;
+	} else if (r == -ERANGE || r == -EOVERFLOW) {
+	  cout << num << ":  got expected ERANGE or EOVERFLOW ? v: " << src_value.version << std::endl;
+	  v_err = true;
+	} else {
+	  cerr << "Error: oid " << oid << " set_redirect " << oid_tgt << " returned error code "
+	       << r << std::endl;
+	  ceph_abort();
+	}
+      } 
+      if (!v_err) {
+	assert(!version || comp->get_version64() == version);
+	version = comp->get_version64();
+	context->update_object_version(oid, comp->get_version64());
+	context->update_object_redirect_target(oid, oid_tgt);
+	if(!src_value.redirect_target.empty()) {
+	  context->oid_redirect_not_in_use.insert(src_value.redirect_target);
+	  context->oid_redirect_in_use.erase(src_value.redirect_target);
+	}      
+      }
+    } else if (info->id == 1) {
+      cout << num << ":  finishing set_redirect stat to oid " << oid << std::endl;
+      if ((r = comp_stat_read->get_return_value())) {
+	if (!(r == -ENOENT && src_value.deleted())) {
+	  cerr << "Error: oid " << oid << " stat " << " returned error code "
+	       << r << std::endl;
+	}
+      } else {
+	assert(comp_stat_read->get_return_value() == 0);
+	if (!v_err) {
+	  assert(!version || comp_stat_read->get_version64() == version);
+	  version = comp_stat_read->get_version64();
+	} else {
+	  cout << num << ": version is changed to v : " << comp_stat_read->get_version64();
+	  context->update_object_version(oid, comp_stat_read->get_version64());
```

Comment:
If I'm following, the time there is a mismatch is when we IGNORE_REDIRECT to do the set-redirect--that's the only time we care about the redirect version and not the target object version.  Can't we just precede any set-redirect with a stat w/ IGNORE_REDIRECT to get the redirect (vs logical object) version?